### PR TITLE
Enable Orchestrator to set pricing by broadcaster ETH address

### PR DESCRIPTION
**Pull request by eliteprox**
**What does this pull request do? Explain your changes. (required)**
Enables orchestrator to set specific pricePerPixel and pricePerUnit by the broadcaster's ETH address.

**Specific updates (required)**
- Adds new CLI option `-pricePerBroadcaster` to livepeer.go 
- Adds new HTTP handler `/setPriceForBroadcaster` to handlers.go and webserver.go
- Adds new livepeer_cli menu option to set broadcaster prices.
- Updates tests for pricing

**How did you test each of these updates (required)**

All tests were ran on testnet with linux using multiple broadcasters and an orchestrator node compiled with the changes.

- Ran livepeer as orchestrator with `-pricePerUnit 5`, `-pixelsPerUnit 1` and -pricePerBroadcaster set to `{"broadcasters":[{"ethaddress":"0x6E98a769926f55A65e34066E67198f6B68e4B166","priceperunit":0,"pixelsperunit":1},{"ethaddress":"0xF115A3dfd39778A6719Ace6AA77A2aD228e90329","priceperunit":1200,"pixelsperunit":1}]}`

- Ran livepeer as orchestrator with `-pricePerUnit 0`, `-pixelsPerUnit 1` and `-pricePerBroadcaster {"broadcasters":[{"ethaddress":"0x6E98a769926f55A65e34066E67198f6B68e4B166","priceperunit":1200,"pixelsperunit":1}]}`

- Tested the same using a json file `-pricePerBroadcaster broadcasterprices.json`.

- Set prices using `curl -X POST -d "pixelsPerUnit=1&pricePerUnit=9&broadcasterEthAddr=0xEB62188121725A605f76a97791a1C69B4BdB435D" http://127.0.0.1:7920/setPriceForBroadcaster`

- Confirmed correct price sent to broadcaster by verifying payment ticket logs.

- Confirmed no deposit required for broadcaster when broadcaster uses `-maxPricePerUnit 0` and orchestrator sets broadcaster's ETH address to a pricePerUnit of 0.

- Confirmed orchestrator's `-pricePerUnit` and `-pricePerPixel` continue to function as the default price when a broadcaster is not assigned a specific price via `-pricePerBroadcaster`, `-freeStream` or `/setPricePerBroadcaster`.

**Does this pull request close any open issues?**
- Implements #2486 

**Checklist:**

- [ x] Read the [contribution guide](./doc/contributing.md)
- [ x] `make` runs successfully
- [ x] All tests in `./test.sh` pass
- [ x] README and other documentation updated
- [ x] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

**Comment by leszko** [comment]
Thank you for the PR! We'll review it in the next 1-2 weeks.

File: CHANGELOG_PENDING.md [hunk 1]
```diff
@@ -1,20 +1,5 @@
-# Unreleased Changes
-
-## vX.X
-
 ### Breaking Changes 🚨🚨
-
```
Comment (leszko) [inline CHANGELOG_PENDING.md]: Please don't remove these parts.

File: cmd/livepeer/livepeer.go [hunk 2]
```diff
@@ -175,6 +175,9 @@ func parseLivepeerConfig() starter.LivepeerConfig {
 	cfg.OrchWebhookURL = flag.String("orchWebhookUrl", *cfg.OrchWebhookURL, "Orchestrator discovery callback URL")
 	cfg.DetectionWebhookURL = flag.String("detectionWebhookUrl", *cfg.DetectionWebhookURL, "(Experimental) Detection results callback URL")
 
+	cfg.PricePerBroadcaster = flag.String("pricePerBroadcaster", *cfg.PricePerBroadcaster, `json list of price per broadcaster. Example: {"broadcasters":[{"ethaddress":"address1","priceperunit":1000,"pixelsperunit":1},{"ethaddress":"address2","priceperunit":1200,"pixelsperunit":1}]}`)
+	cfg.FreeStream = flag.String("freeStream", *cfg.FreeStream, "eth address for broadcaster to provide free stream to")
```
Comment (leszko) [inline cmd/livepeer/livepeer.go]: Do we need this one? I think it should be possible to achieve the same with the proper json in `PricePerBroadcaster`. If that's the case, I'd remove this one, since it's better to limit the number of flags we have in the `livepeer` tool.

File: cmd/livepeer/starter/starter.go [hunk 3]
```diff
@@ -1351,3 +1373,31 @@ func checkOrStoreChainID(dbh *common.DB, chainID *big.Int) error {
 
 	return nil
 }
+
+type setBroadcasterPrice struct {
```
Comment (leszko) [inline cmd/livepeer/starter/starter.go]: Nit: maybe just name it `broadcasterPrice`?

File: cmd/livepeer/starter/starter.go [hunk 4]
```diff
@@ -1351,3 +1373,31 @@ func checkOrStoreChainID(dbh *common.DB, chainID *big.Int) error {
...
+	EthAddress string
+	Price      *big.Rat
+}
+
+func getBroadcasterPrices(broadcasterPrices string) []setBroadcasterPrice {
+	var pricesSet core.BroadcasterPrices
+	prices, _ := common.GetPass(broadcasterPrices)
```
Comment (leszko) [inline cmd/livepeer/starter/starter.go]: Maybe rename `common.GetPass` to something more generic, like `GetOrReadFromFile` or something similar.

File: core/orchestrator.go [hunk 5]
```diff
@@ -51,6 +51,18 @@ type orchestrator struct {
 	secret  []byte
 }
 
+//Format of broadcasterPrices json
```
Comment (leszko) [inline core/orchestrator.go]: If it's used only in `starter.go`, then could you maybe move it there?

File: cmd/livepeer_cli/wizard_transcoder.go [hunk 6]
```diff
@@ -305,3 +305,36 @@ func (w *wizard) setMaxFaceValue() {
 		return
 	}
 }
+
+func (w *wizard) setPriceForBroadcaster() {
+	fmt.Println("Enter the eth address of the broadcaster")
```
Comment (leszko) [inline cmd/livepeer_cli/wizard_transcoder.go]:
```suggestion
	fmt.Println("Enter the ETH address of the broadcaster")
```

File: server/mediaserver.go [hunk 7]
```diff
@@ -1611,6 +1612,8 @@ func (s *LivepeerServer) GetNodeStatus() *common.NodeStatus {
 		OrchestratorPool:      []string{},
 		RegisteredTranscoders: []common.RemoteTranscoderInfo{},
 		LocalTranscoding:      s.LivepeerNode.TranscoderManager == nil,
+		BroadcasterPrices:     make(map[string]*big.Rat),
+		NodeType:              s.LivepeerNode.NodeType.String(),
```
Comment (leszko) [inline server/mediaserver.go]: Do you need this `NodeType` here?

File: common/types.go [hunk 8]
```diff
@@ -36,6 +36,8 @@ type NodeStatus struct {
 	RegisteredTranscodersNumber int
 	RegisteredTranscoders       []RemoteTranscoderInfo
 	LocalTranscoding            bool // Indicates orchestrator that is also transcoder
+	BroadcasterPrices           map[string]*big.Rat
+	NodeType                    string
```
Comment (leszko) [inline common/types.go]: Do you need this `NodeType` here?

File: server/mediaserver.go [hunk 9]
```diff
@@ -1626,6 +1629,9 @@ func (s *LivepeerServer) GetNodeStatus() *common.NodeStatus {
 			res.OrchestratorPool = append(res.OrchestratorPool, info.URL.String())
 		}
 	}
+	if s.LivepeerNode.NodeType.String() == "orchestrator" {
```
Comment (leszko) [inline server/mediaserver.go]: Don't we want to display it in `livepeer_cli`?

**Review by leszko — CHANGES_REQUESTED** [review]
Thank you for the PR. Added a few comments.

File: cmd/livepeer/livepeer.go (hunk 2 above)
Comment (eliteprox) [reply cmd/livepeer/livepeer.go]: Will discuss design with @0xb79

File: CHANGELOG_PENDING.md (hunk 1 above)
Comment (eliteprox) [reply CHANGELOG_PENDING.md]: Restored headings

File: cmd/livepeer/starter/starter.go (hunk 3 above)
Comment (eliteprox) [reply cmd/livepeer/starter/starter.go]: Renamed to broadcasterPrice

File: cmd/livepeer/starter/starter.go (hunk 4 above)
Comment (eliteprox) [reply cmd/livepeer/starter/starter.go]: Renamed GetPass to ReadFromFile

File: cmd/livepeer/livepeer.go (hunk 2 above)
Comment (eliteprox) [reply cmd/livepeer/livepeer.go]: Removed `-FreeStream` flag

File: core/orchestrator.go (hunk 5 above)
Comment (eliteprox) [reply core/orchestrator.go]: Refactored and moved

File: server/mediaserver.go (hunk 7 above)
Comment (eliteprox) [reply server/mediaserver.go]: Removed NodeType

File: server/mediaserver.go (hunk 9 above)
Comment (eliteprox) [reply server/mediaserver.go]: Removed if statement

File: CHANGELOG_PENDING.md [hunk 10]
```diff
@@ -6,18 +6,6 @@
 
 ### Features ⚒
 
-#### General
```
Comment (leszko) [inline CHANGELOG_PENDING.md]: I still see these removed lined. Could you please restore it?

File: cmd/livepeer/livepeer.go [hunk 11]
```diff
@@ -175,6 +175,8 @@ func parseLivepeerConfig() starter.LivepeerConfig {
 	cfg.OrchWebhookURL = flag.String("orchWebhookUrl", *cfg.OrchWebhookURL, "Orchestrator discovery callback URL")
 	cfg.DetectionWebhookURL = flag.String("detectionWebhookUrl", *cfg.DetectionWebhookURL, "(Experimental) Detection results callback URL")
 
+	cfg.PricePerBroadcaster = flag.String("pricePerBroadcaster", *cfg.PricePerBroadcaster, `json list of price per broadcaster. Example: {"broadcasters":[{"ethaddress":"address1","priceperunit":1000,"pixelsperunit":1},{"ethaddress":"address2","priceperunit":1200,"pixelsperunit":1}]}`)
+
```
Comment (leszko) [inline cmd/livepeer/livepeer.go]: Super nit: could you move this line under the line `cfg.AutoAdjustPrice`?

File: cmd/livepeer/starter/starter.go [hunk 12]
```diff
@@ -123,6 +124,7 @@ type LivepeerConfig struct {
 	AuthWebhookURL               *string
 	OrchWebhookURL               *string
 	DetectionWebhookURL          *string
+	PricePerBroadcaster          *string
```
Comment (leszko) [inline cmd/livepeer/starter/starter.go]: Super nit: Could you move this line under `AutoAdjustPrice`?

File: cmd/livepeer/starter/starter.go [hunk 13]
```diff
@@ -196,6 +198,8 @@ func DefaultLivepeerConfig() LivepeerConfig {
 	defaultOrchWebhookURL := ""
 	defaultDetectionWebhookURL := ""
 
+	defaultpricePerBroadcaster := ""
```
Comment (leszko) [inline cmd/livepeer/starter/starter.go]: Super nit: could you move this line under `defaultAutoAdjustPrice`?

File: cmd/livepeer/starter/starter.go [hunk 14]
```diff
@@ -265,6 +269,8 @@ func DefaultLivepeerConfig() LivepeerConfig {
 		AuthWebhookURL:      &defaultAuthWebhookURL,
 		OrchWebhookURL:      &defaultOrchWebhookURL,
 		DetectionWebhookURL: &defaultDetectionWebhookURL,
+
+		PricePerBroadcaster: &defaultpricePerBroadcaster,
```
Comment (leszko) [inline cmd/livepeer/starter/starter.go]: Super nit: could you move this line under `AutoAdjustPrice`?

**Review by leszko — CHANGES_REQUESTED** [review]
Added some super minor comments. Other than that, could you address the following 2 points?

1. The [following GH Action](https://github.com/livepeer/go-livepeer/actions/runs/3155409577/jobs/5141365984) is failing, could you check?

2. I've played a little bit with your change and set the price per broadcaster, but then I don't see it in the status when checked from `livepeer_cli`. Also when I do the following curl call, I don't see the broadcaster prices. Could you check?

```
$ curl localhost:1936/status
{"Manifests":{},"InternalManifests":{},"StreamInfo":{},"OrchestratorPool":[],"OrchestratorPoolInfos":null,"Version":"0.5.34-129325e0","GolangRuntimeVersion":"go1.18.2","GOArch":"arm64","GOOS":"darwin","RegisteredTranscodersNumber":0,"RegisteredTranscoders":[],"LocalTranscoding":true,"BroadcasterPrices":{}}
```

File: server/handlers.go [hunk 15]
```diff
@@ -491,8 +491,13 @@
 		return fmt.Errorf("pixels per unit must be greater than 0, provided %d", pixelsPerUnit)
 	}
 
-	s.LivepeerNode.SetBasePrice(big.NewRat(pricePerUnit, pixelsPerUnit))
-	glog.Infof("Price per pixel set to %d wei for %d pixels\n", pricePerUnit, pixelsPerUnit)
+	s.LivepeerNode.SetBasePrice(broadcasterEthAddr, big.NewRat(pricePerUnit, pixelsPerUnit))
+	if broadcasterEthAddr == "default" {
+		glog.Infof("Price per pixel set to %d wei for %d pixels\n", pricePerUnit, pixelsPerUnit)
+	} else {
+		glog.Infof("Price per pixel set to %d wei for %d pixels for broadcaster %s\n", pricePerUnit, pixelsPerUnit, broadcasterEthAddr)
```
[the remark this answers is not in the document]
Comment (eliteprox) [reply server/handlers.go:506]: Resolved by regex validation

File: CHANGELOG_PENDING.md (hunk 10 above)
Comment (eliteprox) [reply CHANGELOG_PENDING.md]: Restored the missing lines. I has mistakenly thought the existing ones were to be replaced rather than added to the bottom.

File: cmd/livepeer/livepeer.go (hunk 11 above)
Comment (eliteprox) [reply cmd/livepeer/livepeer.go]: Moved to ln 149

File: cmd/livepeer/starter/starter.go (hunk 14 above)
Comment (eliteprox) [reply cmd/livepeer/starter/starter.go]: Moved under AutoAdjustPrice line 249

File: cmd/livepeer/starter/starter.go (hunk 13 above)
Comment (eliteprox) [reply cmd/livepeer/starter/starter.go]: Moved under `defaultAutoAdjustPrice` line 177

File: cmd/livepeer/starter/starter.go (hunk 12 above)
Comment (eliteprox) [reply cmd/livepeer/starter/starter.go]: Moved to line 109 under AutoAdjustPrice

**Comment by leszko** [comment]
@eliteprox Please re-request review or just comment here when you think that your PR is ready for review.

**Comment by leszko** [comment]
btw. I've seen the GH Action started to pass 🎉

**Comment by ad-astra-video** [comment]
@leszko We are adding some tests so the CodeCov reports pass. We had some tests when submitting but it wants a few more.

**Comment by eliteprox** [comment]
> Added some super minor comments. Other than that, could you address the following 2 points?
> 
>     1. The [following GH Action](https://github.com/livepeer/go-livepeer/actions/runs/3155409577/jobs/5141365984) is failing, could you check?
> 
>     2. I've played a little bit with your change and set the price per broadcaster, but then I don't see it in the status when checked from `livepeer_cli`. Also when I do the following curl call, I don't see the broadcaster prices. Could you check?
> 
> 
> ```
> $ curl localhost:1936/status
> {"Manifests":{},"InternalManifests":{},"StreamInfo":{},"OrchestratorPool":[],"OrchestratorPoolInfos":null,"Version":"0.5.34-129325e0","GolangRuntimeVersion":"go1.18.2","GOArch":"arm64","GOOS":"darwin","RegisteredTranscodersNumber":0,"RegisteredTranscoders":[],"LocalTranscoding":true,"BroadcasterPrices":{}}
> ```

Tests are passing and this now works:
`curl localhost:7920/status`

`{"Manifests":{},"InternalManifests":{},"StreamInfo":{},"OrchestratorPool":[],"OrchestratorPoolInfos":null,"Version":"undefined","GolangRuntimeVersion":"go1.18.1","GOArch":"amd64","GOOS":"linux","RegisteredTranscodersNumber":0,"RegisteredTranscoders":[],"LocalTranscoding":false,"BroadcasterPrices":{"0xeb62188121725a605f76a97791a1c69b4bdb435d":"9","default":"2"}`

File: server/mediaserver_test.go [hunk 16]
```diff
@@ -37,10 +37,10 @@ import (
 
 var pushResetWg sync.WaitGroup // needed to synchronize exits from HTTP push
 
-// func setupServer() *LivepeerServer {
-// 	s, _ := setupServerWithCancel()
-// 	return s
-// }
+//	func setupServer() *LivepeerServer {
```
Comment (leszko) [inline server/mediaserver_test.go]: Nit: Please revert this change, it's not related to your PR.

**Review by leszko — COMMENTED** [review]
Revert unrelated change

**Review by leszko — COMMENTED** [review]
Add new line at the end of starter_test.go

**Review by leszko — APPROVED** [review]
LGTM 👍

**Comment by leszko** [comment]
Thanks a lot for this PR @eliteprox
